### PR TITLE
Send data in batch format

### PR DESCRIPTION
**Pull request by samdozor**
This updates the AloomaExtension class to upload data in a single JSON batch, the reasoning is:

- sending data in this format will more closely resemble mParticle's documented JSON models, rather than firehose-specific implementation objects
- should be more efficient to send as a single upload, rather than a bunch of uploads with duplicative data
- this also remove account information from the upload

### CODE REVIEW

File: alooma-extension/src/main/java/com/mparticle/ext/alooma/AloomaExtension.java [hunk 1]
```diff
@@ -204,9 +133,9 @@ public AudienceSubscriptionResponse processAudienceSubscriptionRequest(AudienceS
         return super.processAudienceSubscriptionRequest(request);
     }
 
-    private void sendEvent(Event event) throws IOException{
+    private void sendEvent(EventProcessingRequest eventRequest) throws IOException{
         Response response = webTarget.path("rest/"+token).request()
```
Comment (itamarwe) [inline alooma-extension/src/main/java/com/mparticle/ext/alooma/AloomaExtension.java]: I think it's better to derive the token from the request itself. Avoiding concurrency issues.
@samdozor

Comment (samdozor) [inline alooma-extension/src/main/java/com/mparticle/ext/alooma/AloomaExtension.java]: agreed - though this PR doesn't change that. That's how it is in currently in master.

**Comment by itamarwe** [comment]
The only hesitation I have about this change, is that it will require the user to split the event with some custom code on our end. It's not a big deal, but may cause some friction.

Here's a sample code that a user would have to write:
```python
def transform(event):
  result = []
  if 'events' in event:
    for user_event in event['events']:
      enriched_user_event = user_event
      for field in event:
        if field!='events':
          enriched_user_event[field] = event[field]
      enriched_user_event['_metadata']['event_type'] = 'mparticle'
      result.append(enriched_user_event)
    return result
  return event
```

**Comment by samdozor** [comment]
@itamarwe so to be clear, typically with other integrations, only 1 event is received at a time? If you're fine with all the HTTP overhead and duplicated device info, environment info, etc that's going to be the body of the POST request, that works for me. 

If you agree, I think then the way to go would be to basically serialize a request as i've done, but with just one event each in the request. Still slightly different than the original version, which was serializing the event itself.

**Comment by itamarwe** [comment]
@samdozor Yes, usually we receive 1 event at a time. I think that integration simplicity is more important than the "efficiency" at the moment.

How would you implement your suggestion? We can definitely go ahead with it.

**Comment by samdozor** [comment]
@itamarwe I was thinking something like this:

```
private void sendEvent(EventProcessingRequest eventRequest, String hostname, String token) throws IOException{
        WebTarget webTarget = client.target("https://" + hostname + ".alooma.io").path("rest/"+token);
        List<Event> eventList = eventRequest.getEvents();
        eventRequest.setEvents(new ArrayList<Event>(1));
        for (Event event : eventList) {
            eventRequest.getEvents().add(event);
            Response response = webTarget.request()
                    .accept(MediaType.APPLICATION_JSON_TYPE)
                    .post(Entity.json(eventRequest));
            eventRequest.getEvents().clear();
        }
 }
```

**Comment by itamarwe** [comment]
That could work. I'll test that later.
Thanks.

**Comment by itamarwe** [comment]
Hi Sam,
I tested this solution a little bit, and I think I would still like to stick with our original solution. Sending the raw request makes handling the data less elegant (and thus with more friction for the user) on Alooma's end:
1. The actual event is wrapped in an array, and has to be transformed.
2. The account arrives as a null value, and should be cleared as well.
3. The type is always `EVENT_PROCESSING_REQUEST` and it's not very informative.

So I suggest to keep the code as is on `master`. I find it to be the most elegant end-to-end. I don't mind that the user gets the account details in the event. In any case we add the token in the event metadata, and the host is not sensitive anyway.

Would that work?

Thanks,
Itamar.

**Comment by samdozor** [comment]
Yep, sounds good, we'll go with master. My two concerns to keep an eye on are:

- Large overhead given sending individual requests per event, possibly in the billions per day depending on client.
- mParticle will work on some better documentation around the serialized-firehose API format

**Comment by itamarwe** [comment]
Great. If there is an actual potential client with such high scale who's interested in integrating Alooma we have a couple of roads to take. One is to set up more front-end servers to handle the high load. Another one is to set up an endpoint that can receive batches. But I think it's a discussion worth having once it becomes more concrete.
We are really looking forward to seeing a customer using mParticle-Alooma integration.

**Pull request closed without merging**